### PR TITLE
Enhance kotlin coroutine plugin for stack tracing

### PR DESCRIPTION
**Pull request by devkanro**
# What's changed?
- A new instrumentation point for DispatchedTask
- Provide stack tracing info for kotlin coroutine

## Instrumentation point at `DispatchedTask.run`
In the previous practice, I chose to wrap all `Runnables` as `TracingRunnable`.

This will cause the `CoroutineContext` in `DispachedContinution` to not be used by Dispacher.

Although there will be no problem under normal usage, in some custom `CoroutineContext` scenarios, there may be problems where the `CoroutineContext` cannot be read.

## Provide stack tracing info for kotlin coroutine
Since we can distinguish between a normal `Runnable` and a `DispatchedTask`, we can extract the stack trace of the Kotlin Coroutine suspend from `DispatchedTask`, which may be very effective in actual business.

截屏2023-02-06 18 23 20

# TODO List
- [x] Update changelog
- [ ] Discuss the best way to present the stack
  Since Span's Tag cannot display multiple lines of text, I used multiple tags similar to `coroutine.stack[13]` to display the stack trace. I don't know if it will cause SK performance or other problems.

### CODE REVIEW

**Comment by wu-sheng** [comment]
Why do you need these tags rather than logs directly?

**Comment by devkanro** [comment]
I didn't know it was possible to use log in span, let me try it out

**Comment by wu-sheng** [comment]
When would pick stack when error status is still false? It would a heavily payload in the runtime.
Java agent should be very careful about extra payload.
This seems a concern to me.

**Comment by devkanro** [comment]
The original idea is that because coroutine will switch between threads, if only the stack information of the last exception occurrence is captured, a large amount of stack information may be lost.

But you are right. Recording all of them will cause a heavy payload. The best way would be to only show the coroutine stack in the failed span, don't know if there is a way to do this.

**Comment by wu-sheng** [comment]
I am not sure about your case. Could you draw a graph to show what stacks you want to collect?

**Comment by devkanro** [comment]
截屏2023-02-07 11 10 22
I modified my implementation with two changes:

1. Using log to record stack
2. Just record first user caller frame stack

**Comment by devkanro** [comment]
Next, I will provide config to disable this feature.

**Comment by wu-sheng** [comment]
> 2.Just record first user caller frame stack

I am afraid this doesn't fit upstream requirement. We never have a concept of calling first call.

**Comment by devkanro** [comment]
Perhaps another word is more appropriate, 'the suspend function that occurs in coroutine'

**Comment by wu-sheng** [comment]
I am not sure what do you mean. Could you explain when you try to collecr logs? Why does this help?

**Comment by devkanro** [comment]
When you're viewing a kotlin coroutine's traceback, there are frequent forks across threads.

Knowing the suspend point of each fork can be of great help when problems arise.

test

**Comment by devkanro** [comment]
> When you try to collecr logs? 
When a suspend function is suspended, and submit tasks on other threads.

> Why does this help?
We can know which suspend function is suspended.

**Comment by wu-sheng** [comment]
But stack collecting means stop the current thread running from JVM perspective. This is why the Java best practice doesn't recommend use Exception(stack in exception) to control workflow.
More frequently this happens, more you should not read the stack.

Please be clear why you need this for every request. AFAIK, this would reduce your throughput a lot.

**Comment by devkanro** [comment]
These StackTraceElement come from Kotlin Coroutine object, I will check if source is temporary collecting.

**Comment by wu-sheng** [comment]
I think that is the only way to read real time stack in the JVM.

**Comment by devkanro** [comment]
Kotlin Coroutine will create those StackTraceElement by two ways.

1. Someone generated from compile-time, it can get the most superficial stack, which has no performance cost.  
  [ContinuationImpl.kt#L76](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/jvm/src/kotlin/coroutines/jvm/internal/ContinuationImpl.kt#L76) [DebugMetadata#L41-L49](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/jvm/src/kotlin/coroutines/jvm/internal/DebugMetadata.kt#L41-L49)
2. Another converted from an exception stack trace, these can cause the performance issues you're worried about.
  [DebugProbesImpl#L496](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt#L496)

By default, kotlin provides the second option for every coroutine created, we can refer to the following [documents](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-debug/) to turn off this behavior. 

I plan to only use the compile-time generated StackTraceElement provided in option1 as the source of suspend point logs, so we can get the best performance.

**Comment by wu-sheng** [comment]
Opt1 is just better than opt2, but still have clear cost from extra codes.
I still feel you should only collect stacks in certain conditionss, could you update codes and point to me about this?

File: apm-sniffer/optional-plugins/kotlin-coroutine-plugin/src/main/java/org/apache/skywalking/apm/plugin/kotlin/coroutine/Utils.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import java.util.ArrayList;
+
+public class Utils {
+    private static final Class<?> DISPATCHED_TASK_CLASS;
+
+    static {
+        Class<?> clazz = null;
+        try {
+            clazz = Class.forName(DispatchedTaskInstrumentation.ENHANCE_CLASS);
```
[the remark this answers is not in the document]
Comment (devkanro) [reply apm-sniffer/optional-plugins/kotlin-coroutine-plugin/src/main/java/org/apache/skywalking/apm/plugin/kotlin/coroutine/Utils.java:32]: DispatchedTask is a internal class, I can't access it directly, I will move it to method.

**Comment by wu-sheng** [comment]
We need to
1. Vocabulary doc for new config
2. Add new config to agent.config
3. Update the your breach updated to main branch. I can't do sync due to your setup.

**Comment by devkanro** [comment]
> I still feel you should only collect stacks in certain conditionss, could you update codes and point to me about this?

Theoretically speaking, it should be collected when an exception occurs, but the span that has already occurred cannot be notified, so I use the config method to switch this function here, and the default is off.

**Comment by wu-sheng** [comment]
Span is not that heavy, only errorb log is. Could you collect logs(despite span) when error happens only?

**Comment by devkanro** [comment]
How can I add my newly added configuration to the default agent.conf?

**Comment by wu-sheng** [comment]
What do you mean how? This is a config file only.

**Comment by devkanro** [comment]
The `package` task will generate a default agent.config to `skywalking-agent/config/agent.config`, I thought it would be auto-generated from @PluginConfig, but it looks like KotlinCoroutinePluginConfig fields don't show up in it.

**Comment by devkanro** [comment]
Maybe I can collect stack info for the coroutine that went wrong, let me try it out

**Comment by devkanro** [comment]
It works with some tricks.

Kotlin Coroutine span will log the error.
截屏2023-02-08 15 14 52

With error stack trace.
截屏2023-02-08 15 15 29

And coroutine stack
截屏2023-02-08 15 15 51

File: apm-sniffer/optional-plugins/kotlin-coroutine-plugin/src/main/java/org/apache/skywalking/apm/plugin/kotlin/coroutine/Utils.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+public class Utils {
+    private static Class<?> DISPATCHED_TASK_CLASS;
+
+    public static boolean isDispatchedTask(Runnable runnable) {
+        if (DISPATCHED_TASK_CLASS == null) {
+            try {
+                DISPATCHED_TASK_CLASS = Class.forName(DispatchedTaskInstrumentation.ENHANCE_CLASS);
+            } catch (ClassNotFoundException ignored) {
+            }
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/kotlin-coroutine-plugin/src/main/java/org/apache/skywalking/apm/plugin/kotlin/coroutine/Utils.java:34]: Currently, these codes seem going to be endless, if the ClassNotFoundException really happens. Is there any chance this happens? Or this is just for adopting Java Exception catch compiling requirement.

**Comment by wu-sheng** [comment]
Thanks for polishing the plugin patiently. Just one question left.

**Comment by devkanro** [comment]
In fact, when the DispatchedTask does not exist, it will not be executed here, because the Interceptor will not be created.
But I can modify this code to make more sense.

**Comment by wu-sheng** [comment]
That would be better. To use a class existence as condition in a plugin is not recommended usually.

**Review by wu-sheng — APPROVED** [review]
LGTM

**Pull request merged**